src/Plugin/PasswordConstraint/MinimumCharacters.php

```diff
@@ -0,0 +1,102 @@
+<?php
+
+namespace Drupal\password_enhancements\Plugin\PasswordConstraint;
+
+use Drupal\Core\Form\FormStateInterface;
+use Drupal\Core\Plugin\PluginBase;
+
+/**
+ * Defines a common class for minimum character requirement.
+ */
+abstract class MinimumCharacters extends PluginBase {
+
+  /**
+   * The generated error message.
+   *
+   * @var string|null
+   */
+  protected $errorMessage;
+
+  /**
+   * {@inheritdoc}
+   */
+  public function __construct(array $configuration, $plugin_id, $plugin_definition) {
+    parent::__construct($configuration, $plugin_id, $plugin_definition);
+
+    $this->errorMessage = NULL;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getErrorMessage(): ?string {
+    return $this->errorMessage;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getInitialDescription(): string {
+    if ($this->configuration['minimum_characters'] > 1) {
+      return $this->configuration['descriptionPlural'];
+    }
+    else {
+      return $this->configuration['descriptionSingular'];
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function settingsForm(): array {
+    $form = [];
+
+    $form['minimum_characters'] = [
+      '#type' => 'number',
+      '#title' => $this->t('Minimum characters'),
+      '#required' => TRUE,
+      '#min' => 1,
+      '#default_value' => !empty($this->configuration['minimum_characters']) ? $this->configuration['minimum_characters'] : 1,
+    ];
+
+    return $form;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function settingsValidate(array &$form, FormStateInterface $form_state): void {
+    if ($form_state->getValue(['settings', 'minimum_characters']) < 1) {
+      $form_state->setError($form['settings']['minimum_characters'], $this->t('The %title field must be a non-zero, positive number.', [
+        '%title' => $form['settings']['minimum_characters']['#title'],
+      ]));
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function settingsSubmit(array &$form, FormStateInterface $form_state): void {
+    // Nothing to do here.
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function validateValue(string $value, bool $custom_error_message = FALSE): bool {
+    $character_count = mb_strlen($value);
+    if ($this->configuration['minimum_characters'] <= $character_count) {
+      $this->errorMessage = NULL;
+      return TRUE;
+    }
+
+    if (!$custom_error_message) {
```

Comment:
1. What if `$custom_error_message` is TRUE? Where and how do you handle that?
2. Wherever I see this method called in your code (except: `SpecialCharacter::validate()`) you only call this method with _only one parameter_.

Reply:
1. If set to true, the caller function has to handle the error message.
2. Yes, SpecialCharacter requires a custom message.

I've also updated the doxygen for this.

Reply:
> Yes, SpecialCharacter requires a custom message.

So you added a parameter to a method in a parent class which is inherited by all its subclasses just to be able to use it in only one subclass? Well, this is not how you should use inheritance. Sometimes encapsulation is an answer to this question or moving some code to traits. In your case, I think I managed to simplify for your in multiple ways by keeping the inheritance. (Did not run the code, it could break, but you should see what changes I am suggesting and hopefully why. Let me know if you have any questions.)

[refactoring_inharitance_tree_and_exception_handling.diff.zip](https://github.com/Pronovix/password_enhancements/files/3354270/refactoring_inharitance_tree_and_exception_handling.diff.zip)

Reply:
Solved differently with exceptions as you suggested.